src/tracing/esp32_trace/include/matter/tracing/insights_sys_stats.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "insights_sys_stats.h"
+#include <esp_diagnostics_metrics.h>
+#include <esp_err.h>
+#include <esp_log.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <string.h>
+#include <system/SystemStats.h>
+
+namespace chip {
+namespace System {
+namespace Stats {
+
+InsightsSystemMetrics & InsightsSystemMetrics::GetInstance()
+{
+    static InsightsSystemMetrics instance;
+    return instance;
+}
+
+void InsightsSystemMetrics::SamplingHandler(Layer * systemLayer, void * context)
+{
+    auto instance            = static_cast<InsightsSystemMetrics *>(context);
+    count_t * highwatermarks = GetHighWatermarks();
+    for (int i = 0; i < System::Stats::kNumEntries; i++)
+    {
+        esp_err_t err = esp_diag_metrics_add_uint(instance->mLabels[i], static_cast<int>(highwatermarks[i]));
+        if (err != ESP_OK)
+        {
+            ESP_LOGE(kTag, "Failed to add the metric:%s, err:%d", instance->mLabels[i], err);
+        }
+    }
+    DeviceLayer::SystemLayer().StartTimer(instance->GetSamplingInterval(), SamplingHandler, instance);
+}
+
+CHIP_ERROR InsightsSystemMetrics::Unregister()
+{
+    if (!mRegistered)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    for (int i = 0; i < System::Stats::kNumEntries; i++)
+    {
+        if (mLabels[i] != nullptr)
+        {
+            esp_err_t err = esp_diag_metrics_unregister(mLabels[i]);
+            if (err != ESP_OK)
+            {
+                ESP_LOGE(kTag, "Failed to unregister metric:%s, err:%d", mLabels[i], err);
+            }
+            free(mLabels[i]);
+            mLabels[i] = nullptr;
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+void InsightsSystemMetrics::SetSamplingHandler(intptr_t arg)
+{
+    auto instance = static_cast<InsightsSystemMetrics *>(reinterpret_cast<void *>(arg));
```

Comment:
Casting `arg` to `InsightsSystemMetrics` should work.
```suggestion
    auto instance = static_cast<InsightsSystemMetrics *>(arg);
```